# Project 3, Part 2, Queries on the stations, lines, and travel times tables for the BART system that we will use later to create the graph

University of California, Berkeley

Master of Information and Data Science (MIDS) program

w205 - Fundamentals of Data Engineering



# Included Modules and Packages

Code cell containing your includes for modules and packages

Some starter code is provided

You may change the starter code as needed

You may add as much code and/or as many code cells as you need

In [1]:
import csv

import math
import numpy as np
import pandas as pd

import psycopg2

# Supporting code

Code cells containing any supporting code, such as connecting to the database, any functions, etc.  

Remember you can freely use any code from the labs. You do not need to cite code from the labs.

Some starter code is provided

You may change the starter code as needed

You may add as much code and/or as many code cells as you need

In [2]:
#
# function to run a select query and return rows in a pandas dataframe
# pandas puts all numeric values from postgres to float
# if it will fit in an integer, change it to integer
#

def my_select_query_pandas(query, rollback_before_flag, rollback_after_flag):
    "function to run a select query and return rows in a pandas dataframe"
    
    if rollback_before_flag:
        connection.rollback()
    
    df = pd.read_sql_query(query, connection)
    
    if rollback_after_flag:
        connection.rollback()
    
    # fix the float columns that really should be integers
    
    for column in df:
    
        if df[column].dtype == "float64":

            fraction_flag = False

            for value in df[column].values:
                
                if not np.isnan(value):
                    if value - math.floor(value) != 0:
                        fraction_flag = True

            if not fraction_flag:
                df[column] = df[column].astype('Int64')
    
    return(df)
    

In [3]:
connection = psycopg2.connect(
    user = "postgres",
    password = "ucb",
    host = "postgres",
    port = "5432",
    database = "postgres"
)

In [4]:
cursor = connection.cursor()

# 3.2.1 Query the list of stations

Show the station from the stations table

Sort by station

The query should return 50 rows

Display the results in a Pandas data frame

In [5]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select * from stations order by station;

"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

station   latitude   longitude  transfer_time
0            12th Street  37.803608 -122.272006            282
1    16th Street Mission  37.764847 -122.420042            287
2            19th Street  37.807869 -122.268980             67
3    24th Street Mission  37.752000 -122.418700            277
4                Antioch  37.996281 -121.783404              0
5                  Ashby  37.853068 -122.269957            299
6            Balboa Park  37.721667 -122.447500             48
7               Bay Fair  37.697000 -122.126500             63
8              Berryessa  37.368361 -121.874655            288
9          Castro Valley  37.690748 -122.075679              0
10          Civic Center  37.779861 -122.413498            325
11              Coliseum  37.753611 -122.196944             54
12                 Colma  37.684722 -122.466111            276
13               Concord  37.973745 -122.029127              0
14             Daly City  37.706224 -122.468934            285
15     Downtown Berkeley  37.869799 -122.268197            323
16                Dublin  37.701663 -121.899232              0
17  El Cerrito del Norte  37.925183 -122.316939            311
18      El Cerrito Plaza  37.902694 -122.298968            280
19           Embarcadero  37.793056 -122.397222            304
20               Fremont  37.557489 -121.976620            306
21             Fruitvale  37.774800 -122.224100            279
22             Glen Park  37.733118 -122.433808            314
23               Hayward  37.669700 -122.087000            284
24             Lafayette  37.893186 -122.124614              0
25          Lake Merritt  37.797773 -122.266588            309
26             MacArthur  37.828260 -122.267275             59
27              Millbrae  37.600300 -122.386700              0
28              Milpitas  37.410278 -121.891111            292
29     Montgomery Street  37.789355 -122.401942            313
30        North Berkeley  37.873915 -122.282552            289
31         North Concord  38.003273 -122.024599              0
32                   OAK  37.713200 -122.212200              0
33                Orinda  37.878427 -122.183740              0
34             Pittsburg  38.018869 -121.944208              0
35      Pittsburg Center  38.018227 -121.890178              0
36         Pleasant Hill  37.928399 -122.055992              0
37         Powell Street  37.784000 -122.408000            286
38              Richmond  37.936811 -122.353095            302
39             Rockridge  37.844452 -122.252083              0
40             San Bruno  37.638300 -122.416500            307
41           San Leandro  37.721764 -122.160684            298
42                   SFO  37.616400 -122.391000            291
43         South Hayward  37.634362 -122.057172            319
44   South San Francisco  37.664264 -122.444043            316
45            Union City  37.590746 -122.017282            293
46          Walnut Creek  37.905724 -122.067332              0
47          Warm Springs  37.503000 -121.940000            278
48           West Dublin  37.699726 -121.928273              0
49          West Oakland  37.804900 -122.295100            283

# 3.2.2 Query the lines served by the West Oakland station

The West Oakland station should serve the 4 following lines:

|station|line|
|-------|----|
|West Oakland|blue|
|West Oakland|green|
|West Oakland|red|
|West Oakland|yellow|

Show the station and line from the lines table where the station is West Oakland

Sort by station, line

Display the results in a Pandas data frame

In [8]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select station, line from lines where station='West Oakland';

"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

station    line
0  West Oakland    blue
1  West Oakland   green
2  West Oakland     red
3  West Oakland  yellow

# 3.2.3 Query the list of stations and the lines they serve

Same query as the last one, but for all stations, not just limited to West Oakland

There should be 114 row returned

Display the results in a Pandas data frame

In [9]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select station, line from lines;

"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

station    line
0                 Dublin    blue
1            West Dublin    blue
2          Castro Valley    blue
3               Bay Fair    blue
4            San Leandro    blue
..                   ...     ...
109            Daly City  yellow
110                Colma  yellow
111  South San Francisco  yellow
112            San Bruno  yellow
113                  SFO  yellow

[114 rows x 2 columns]

# 3.2.4 Query the list of all possible line transfers at the West Oakland station and the transfer times

A transfer is when two or more lines pass through a station and the passenger gets off one line and gets on another line.

At West Oakland, there are 12 possible transfers:

|station|from_line|to_line|transfer_time|
|-------|---------|-------|-------------|
|West Oakland|blue|green|283|
|West Oakland|blue|red|283|
|West Oakland|blue|yellow|283|
|West Oakland|green|blue|283|
|West Oakland|green|red|283|
|West Oakland|green|yellow|283|
|West Oakland|red|blue|283|
|West Oakland|red|green|283|
|West Oakland|red|yellow|283|
|West Oakland|yellow|blue|283|
|West Oakland|yellow|green|283|
|West Oakland|yellow|red|283|

Show the station, from line, to line, transfer time in seconds where the station is West Oakland

Sort by station, from line, to line

Hints:

Join the lines table to itself where the station is the same, but the lines are different

Then join to the stations table to pull the transfer time

Display the results in a Pandas data frame

In [13]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select 
  l.*,
  s.transfer_time
from (
  select 
    a.station,
    a.line as from_line,
    b.line as to_line
  from lines a, lines b
  where a.station=b.station and a.line!=b.line and a.station='West Oakland'
) l
join stations s
on l.station=s.station
;

"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

station from_line to_line  transfer_time
0   West Oakland      blue   green            283
1   West Oakland      blue     red            283
2   West Oakland      blue  yellow            283
3   West Oakland     green    blue            283
4   West Oakland     green     red            283
5   West Oakland     green  yellow            283
6   West Oakland       red    blue            283
7   West Oakland       red   green            283
8   West Oakland       red  yellow            283
9   West Oakland    yellow    blue            283
10  West Oakland    yellow   green            283
11  West Oakland    yellow     red            283

# 3.2.5 Query the list of all possible line transfers and the transfer times

Same query as the last query, but all stations, not just limited to West Oakland

The query should return 208 rows

Display the results in a Pandas data frame

In [14]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select 
  l.*,
  s.transfer_time
from (
  select 
    a.station,
    a.line as from_line,
    b.line as to_line
  from lines a, lines b
  where a.station=b.station and a.line!=b.line
) l
join stations s
on l.station=s.station
;

"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

station from_line to_line  transfer_time
0     12th Street    yellow     red            282
1     12th Street    yellow  orange            282
2     12th Street       red  yellow            282
3     12th Street       red  orange            282
4     12th Street    orange  yellow            282
..            ...       ...     ...            ...
203  West Oakland     green     red            283
204  West Oakland     green    blue            283
205  West Oakland      blue  yellow            283
206  West Oakland      blue     red            283
207  West Oakland      blue   green            283

[208 rows x 4 columns]

# 3.2.6 Query the list of all segments between the West Oakland station and its adjoining stations


West Oakland has the following 8 segments between it and the stations it adjoins:

|line|from station|to station|travel time in seconds|
|----|------------|----------|----------------------|
|blue|Lake Merritt|West Oakland|360|
|blue|West Oakland|Embarcadero|420|
|green|Lake Merritt|West Oakland|360|
|green|West Oakland|Embarcadero|420|
|red|12th Street|West Oakland|300|
|red|West Oakland|Embarcadero|420|
|yellow|12th Street|West Oakland|300|
|yellow|West Oakland|Embarcadero|420|

Show the line, from station, to station, travel time in seconds

Sort by line, from station, to station

Note: Since the lines table is only showing one way for each line, we will only get one way in this query.  This is fine.  Later when we build the graph, we will simply create the relationships going both ways.

Hints:

Join lines to itself where the line is the same and the sequence is equal to the sequence + 1

Then join to the travel_times table matching stations and pulling the travel time (note that the stations may be in either order in the travel_times table)

Display the results in a Pandas data frame

In [28]:
rollback_before_flag = True
rollback_after_flag = True

query = """

with t1 as (
  select 
    a.line,
    b.station as from_station,
    a.station as to_station
  from lines a, lines b
  where a.line=b.line and a.sequence=b.sequence+1 
)

select 
  a.*,
  b.travel_time as travel_time_in_seconds
from t1 a
join travel_times b
on 
  a.from_station=b.station_1 and a.to_station=b.station_2
  or a.from_station=b.station_2 and a.to_station=b.station_1
where a.from_station='West Oakland' or a.to_station='West Oakland'
order by 1, 4
;

"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

line  from_station    to_station  travel_time_in_seconds
0    blue  Lake Merritt  West Oakland                     360
1    blue  West Oakland   Embarcadero                     420
2   green  Lake Merritt  West Oakland                     360
3   green  West Oakland   Embarcadero                     420
4     red   12th Street  West Oakland                     300
5     red  West Oakland   Embarcadero                     420
6  yellow   12th Street  West Oakland                     300
7  yellow  West Oakland   Embarcadero                     420

# 3.2.7 Query the list of all segments between each station and its adjoining stations

Same query as the last query, but for all stations, not limited to West Oakland.

The query should return 108 rows

Display the results in a Pandas data frame

In [29]:
rollback_before_flag = True
rollback_after_flag = True

query = """

with t1 as (
  select 
    a.line,
    b.station as from_station,
    a.station as to_station
  from lines a, lines b
  where a.line=b.line and a.sequence=b.sequence+1 
)

select 
  a.*,
  b.travel_time as travel_time_in_seconds
from t1 a
join travel_times b
on 
  a.from_station=b.station_1 and a.to_station=b.station_2
  or a.from_station=b.station_2 and a.to_station=b.station_1
order by 1, 4
;

"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

line         from_station           to_station  travel_time_in_seconds
0      blue          Embarcadero    Montgomery Street                      60
1      blue        Powell Street         Civic Center                      60
2      blue    Montgomery Street        Powell Street                     120
3      blue  16th Street Mission  24th Street Mission                     120
4      blue            Glen Park          Balboa Park                     120
..      ...                  ...                  ...                     ...
103  yellow              Concord        Pleasant Hill                     360
104  yellow            Pittsburg        North Concord                     360
105  yellow         West Oakland          Embarcadero                     420
106  yellow              Antioch     Pittsburg Center                     420
107  yellow     Pittsburg Center            Pittsburg                     600

[108 rows x 4 columns]